crimson/osd: make PGRecovery::start_primary_recovery_ops take objs recovered by UrgentRecovery into account

xxhdx1985126 opened the pull request:
this is similar to [cf07974c](https://github.com/ceph/ceph/pull/37943/commits/cf07974c895d4c0d50f77f921cf10846d1cc274f)

Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

xxhdx1985126 commented:
jenkins test make check

athanatos commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -150,6 +150,9 @@ size_t PGRecovery::start_primary_recovery_ops(
 	++skipped;
+	auto& recovery_waiter = pg->get_recovery_backend()->get_recovering(soid);
+	out->push_back(recovery_waiter.wait_for_recovered_blocking());
+	started++;
```
I don't really understand what's going on here, why do we both increment skipped and started?  Why do we only care if we're not recovering soid, but are recovering head?

xxhdx1985126 commented:
@athanatos Sorry, that last commit was wrong, and I've corrected it. The reason that I added this commit is that BackgroundRecovery requests may fall into an infinite start_primary_recovery_ops loop if they have no objects to recover and all inflight recovering objects are recovered by UrgentRecovery requests. I've just added a ticket for this bug: [48868](https://tracker.ceph.com/issues/48868). Please take a look:-) Thanks.

tchaikov commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -154,6 +154,10 @@ size_t PGRecovery::start_primary_recovery_ops(
 	out->push_back(recover_missing(soid, item.need));
 	++started;
       }
+    } else {
```
@xxhdx1985126 how about transpose these two branches and flatten these conditions? so the code reads
```c++
  if (pg->get_recovery_backend()->is_recovering(soid)) {
    auto& recovery_waiter = pg->get_recovery_backend()->get_recovering(soid);
    out->push_back(recovery_waiter.wait_for_recovered_blocking());
    ++started;
  } else if (pg->get_recovery_backend()->is_recovering(head)) {
    ++skipped;
  } else {
    out->push_back(recover_missing(soid, item.need));
    ++started;
  }
 
```

xxhdx1985126 replied:
@tchaikov Done:-)

tchaikov commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -147,13 +147,15 @@ size_t PGRecovery::start_primary_recovery_ops(
+    if (pg->get_recovery_backend()->is_recovering(soid)) {
+      auto& recovery_waiter = pg->get_recovery_backend()->get_recovering(soid);
+      out->push_back(recovery_waiter.wait_for_recovered_blocking());
+      started++;
```
@xxhdx1985126 

how about

```suggestion
      ++started;
```

to be more consistent?

xxhdx1985126 replied:
done:-)

tchaikov replied:
@xxhdx1985126 did you forget to push the updated branch? it still looks the same.

tchaikov commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -147,13 +147,15 @@ size_t PGRecovery::start_primary_recovery_ops(
+      ++skipped;
+    } else {
+      out->push_back(recover_missing(soid, item.need));
+      started++;
```
i think ``++started`` is better. it is consistent with ``++skipped``. and since we don't care about the value before increment, `++started`` is perfectly fine.

xxhdx1985126 replied:
Sorry, I changed the wrong line. Just corrected it.
